### PR TITLE
feat(i18n): pass translation config directly into ngc

### PR DESCRIPTION
**Pull request by matsko**

### CODE REVIEW

File: modules/@angular/compiler-cli/src/compiler_private.ts [hunk 1]
```diff
@@ -53,3 +53,11 @@ export var NgModuleCompiler: typeof _c.NgModuleCompiler = _c.NgModuleCompiler;
 
 export type TypeScriptEmitter = _c.TypeScriptEmitter;
 export var TypeScriptEmitter: typeof _c.TypeScriptEmitter = _c.TypeScriptEmitter;
+
+export type Xtb = _c.Xtb;
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/compiler_private.ts]: all of those should be public ??

File: modules/@angular/compiler-cli/integrationtest/test/basic_spec.ts [hunk 2]
```diff
@@ -44,24 +44,40 @@ describe('template codegen output', () => {
...
   });
+
+  describe('i18n', () => {
+    it('should support i18n for content tags', () => {
+      var compFixture = createComponent(BasicComp);
+      expect(compFixture.componentInstance.ctxLocale).toEqual('FI');
+    });
+
+    it('should support i18n for content tags', () => {
+      var compFixture = createComponent(BasicComp);
```
Comment (vicb) [inline modules/@angular/compiler-cli/integrationtest/test/basic_spec.ts]: const

File: modules/@angular/compiler-cli/integrationtest/test/basic_spec.ts [hunk 3]
```diff
@@ -44,24 +44,40 @@ describe('template codegen output', () => {
...
     compFixture.detectChanges();
-    expect(debugElement.children.length).toBe(4);
+    expect(debugElement.children.length).toBe(5);
     expect(debugElement.children[2].attributes['value']).toBe('1');
     expect(debugElement.children[3].attributes['value']).toBe('2');
   });
+
+  describe('i18n', () => {
+    it('should support i18n for content tags', () => {
+      var compFixture = createComponent(BasicComp);
```
Comment (vicb) [inline modules/@angular/compiler-cli/integrationtest/test/basic_spec.ts]: const

File: modules/@angular/compiler-cli/integrationtest/src/basic.ts [hunk 4]
```diff
@@ -23,5 +23,13 @@ export class BasicComp {
   ctxProp: string;
   ctxBool: boolean;
   ctxArr: any[] = [];
-  constructor() { this.ctxProp = 'initialValue'; }
+  ctxLocale: string;
+  ctxTranslationsFormat: string;
+  constructor(
+      @Inject(LOCALE_ID) localeID: string,
```
Comment (vicb) [inline modules/@angular/compiler-cli/integrationtest/src/basic.ts]: `@Inject(LOCALE_ID) public localeId: string`

File: modules/@angular/compiler-cli/integrationtest/test/basic_spec.ts [hunk 5]
```diff
@@ -44,24 +44,45 @@ describe('template codegen output', () => {
...
-    expect(debugElement.children.length).toBe(4);
+    expect(debugElement.children.length).toBe(5);
     expect(debugElement.children[2].attributes['value']).toBe('1');
     expect(debugElement.children[3].attributes['value']).toBe('2');
   });
+
+  describe('i18n', () => {
+    it('should inject the locale into the component', () => {
+      const compFixture = createComponent(BasicComp);
+      expect(compFixture.componentInstance.ctxLocale).toEqual('FI');
```
Comment (vicb) [inline modules/@angular/compiler-cli/integrationtest/test/basic_spec.ts]: fi

File: modules/@angular/compiler-cli/src/codegen.ts [hunk 6]
```diff
@@ -88,6 +93,35 @@ export class CodeGenerator {
     return path.join(this.options.genDir, relativePath);
   }
 
+  private _constructTranslationBundle(fileMetas: any[], analyzedNgModules: any): Promise<any> {
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/codegen.ts]: remove !

File: modules/@angular/compiler-cli/src/codegen.ts [hunk 7]
```diff
@@ -97,28 +131,37 @@ export class CodeGenerator {
       return ngModules;
     }, <StaticSymbol[]>[]);
     const analyzedNgModules = this.compiler.analyzeModules(ngModules);
-    return Promise
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/codegen.ts]: revert

File: modules/@angular/compiler-cli/src/extract_i18n.ts [hunk 8]
```diff
@@ -28,8 +28,9 @@ import {StaticAndDynamicReflectionCapabilities} from './static_reflection_capabi
 import {StaticReflector, StaticSymbol} from './static_reflector';
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/extract_i18n.ts]: revert this file

File: modules/@angular/compiler-cli/src/codegen.ts [hunk 9]
```diff
@@ -32,7 +35,9 @@ const PREAMBLE = `/**
 
 export class CodeGenerator {
   constructor(
-      private options: AngularCompilerOptions, private program: ts.Program,
+      private options: AngularCompilerOptions, private transContent: string,
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/codegen.ts]: append at the end to make diff readable

File: modules/@angular/compiler-cli/src/codegen.ts [hunk 10]
```diff
@@ -132,7 +175,7 @@ export class CodeGenerator {
     const reflectorHost = new ReflectorHost(program, compilerHost, options, reflectorHostContext);
     const staticReflector = new StaticReflector(reflectorHost);
     StaticAndDynamicReflectionCapabilities.install(staticReflector);
-    const htmlParser = new compiler.i18n.HtmlParser(new HtmlParser());
+    const htmlParser = new compiler.i18n.HtmlParser(new HtmlParser(), transContent);
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/codegen.ts]: This is the only line that you should keep beside loading the file

File: modules/@angular/compiler-cli/integrationtest/src/basic.ts [hunk 11]
```diff
@@ -23,5 +23,9 @@ export class BasicComp {
   ctxProp: string;
   ctxBool: boolean;
   ctxArr: any[] = [];
-  constructor() { this.ctxProp = 'initialValue'; }
+  constructor(
+      @Inject(LOCALE_ID) public localeID: string,
```
Comment (vicb) [inline modules/@angular/compiler-cli/integrationtest/src/basic.ts]: Id

File: modules/@angular/compiler-cli/src/extract_i18n.ts [hunk 12]
```diff
@@ -28,8 +28,9 @@ import {StaticAndDynamicReflectionCapabilities} from './static_reflection_capabi
...
 
 function extract(
-    ngOptions: tsc.AngularCompilerOptions, program: ts.Program, host: ts.CompilerHost) {
-  const extractor = Extractor.create(ngOptions, program, host);
+    ngOptions: tsc.AngularCompilerOptions, cliOptions: tsc.CliOptions, program: ts.Program,
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/extract_i18n.ts]: only add the format

File: modules/@angular/compiler-cli/src/extract_i18n.ts [hunk 13]
```diff
@@ -142,7 +143,7 @@ export class Extractor {
     const reflectorHost = new ReflectorHost(program, compilerHost, options, reflectorHostContext);
     const staticReflector = new StaticReflector(reflectorHost);
     StaticAndDynamicReflectionCapabilities.install(staticReflector);
-    const htmlParser = new HtmlParser();
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/extract_i18n.ts]: revert

File: modules/@angular/compiler-cli/src/extract_i18n.ts [hunk 14]
```diff
@@ -163,7 +164,7 @@ export class Extractor {
         config, console, elementSchemaRegistry, staticReflector);
     const offlineCompiler = new compiler.OfflineCompiler(
         resolver, normalizer, tmplParser, new StyleCompiler(urlResolver), new ViewCompiler(config),
-        new NgModuleCompiler(), new TypeScriptEmitter(reflectorHost));
+        new NgModuleCompiler(), new TypeScriptEmitter(reflectorHost), '', translationsFormat);
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/extract_i18n.ts]: `, null, null` would be more semantically correct

File: modules/@angular/compiler-cli/integrationtest/test/basic_spec.ts [hunk 15]
```diff
@@ -44,24 +44,45 @@ describe('template codegen output', () => {
...
+
+    it('should inject the translations format into the component', () => {
+      const compFixture = createComponent(BasicComp);
+      expect(compFixture.componentInstance.translationsFormat).toEqual('xtb');
+    });
+
+    it('should support i18n for content tags', () => {
+      const compFixture = createComponent(BasicComp);
+      const debugElement = compFixture.debugElement;
+      const containerElement = <any>debugElement.nativeElement;
```
Comment (vicb) [inline modules/@angular/compiler-cli/integrationtest/test/basic_spec.ts]: `expect dE.query(By.css(sel)).toHaveText('blablabla');`

Comment (matsko) [inline modules/@angular/compiler-cli/integrationtest/test/basic_spec.ts]: This isn't a protractor test.

Comment (vicb) [inline modules/@angular/compiler-cli/integrationtest/test/basic_spec.ts]: it should still work... see `i18n/integration_spec.ts`

File: modules/@angular/compiler-cli/src/extract_i18n.ts [hunk 16]
```diff
@@ -183,11 +184,15 @@ interface FileMetadata {
 // Entry point
 if (require.main === module) {
   const args = require('minimist')(process.argv.slice(2));
-  tsc.main(args.p || args.project || '.', args.basePath, extract)
-      .then(exitCode => process.exit(exitCode))
-      .catch(e => {
+  const project = args.p || args.project || '.';
+  // we have an empty object here because ng-xi18n does
+  // not currently have an cli input arguments
+  const options = new tsc.CliOptions({});
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/extract_i18n.ts]: it has the format

File: modules/@angular/compiler-cli/src/codegen.ts [hunk 17]
```diff
@@ -22,6 +22,8 @@ import {GENERATED_FILES, ReflectorHost, ReflectorHostContext} from './reflector_
 import {StaticAndDynamicReflectionCapabilities} from './static_reflection_capabilities';
 import {StaticReflector, StaticSymbol} from './static_reflector';
 
+var nodeFs = require('fs');
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/codegen.ts]: const

File: modules/@angular/compiler-cli/src/codegen.ts [hunk 18]
```diff
@@ -117,8 +119,15 @@ export class CodeGenerator {
   }
 
   static create(
-      options: AngularCompilerOptions, program: ts.Program, compilerHost: ts.CompilerHost,
-      reflectorHostContext?: ReflectorHostContext): CodeGenerator {
+      options: AngularCompilerOptions, cliOptions: CliOptions, program: ts.Program,
+      compilerHost: ts.CompilerHost, reflectorHostContext?: ReflectorHostContext): CodeGenerator {
+    const transFile = cliOptions.i18nFile;
+    const locale = cliOptions.locale;
+    var transContent: string = '';
```
Comment (vicb) [inline modules/@angular/compiler-cli/src/codegen.ts]: let (var should be banned)

**Comment by vicb** [comment]
Please address all the remaining comments & rebase on master

File: modules/@angular/core/index.ts [hunk 19]
```diff
@@ -25,6 +25,7 @@ export * from './src/testability/testability';
 export * from './src/change_detection';
 export * from './src/platform_directives_and_pipes';
 export * from './src/platform_core_providers';
+export * from './src/i18n/tokens';
```
Comment (vicb) [inline modules/@angular/core/index.ts]: best practice is to list the export explicitly (even if others don't)

File: tools/@angular/tsc-wrapped/src/main.ts [hunk 20]
```diff
@@ -6,19 +6,22 @@ import {check, tsc} from './tsc';
...
+    project: string, cliOptions: CliOptions, codegen?: CodegenExtension): Promise<any> {
   try {
     let projectDir = project;
     if (fs.lstatSync(project).isFile()) {
       projectDir = path.dirname(project);
     }
 
     // file names in tsconfig are resolved relative to this absolute path
-    basePath = path.resolve(process.cwd(), basePath || projectDir);
+    var basePath = path.resolve(process.cwd(), cliOptions.basePath || projectDir);
```
Comment (vicb) [inline tools/@angular/tsc-wrapped/src/main.ts]: const

**Comment by alexeagle** [comment]
please CC me on `ngc` changes.
We're trying to get it into google3, and I am not sure about this extra configuration outside of tsconfig.json
it makes all our tooling harder to have some options in a param file and some not.
@vicb can we discuss??

**Comment by vicb** [comment]
Aloha, the idea is to invoke NGC once per locale which is why the locals can not be in the tsconfig

**Comment by alexeagle** [comment]
I disagree pretty strongly with having command-line arguments to ngc differ from tsc.
1. can't swap `ngc` with `tsc` on the command line to simplify a problem
2. no path to upstream this into typescript
3. a team has to document the command to run since the `tsconfig.json` is not sufficient
4. similarly, no tools understand what languages will be produced because the configuration is not in the config file
5. it makes the tooling harder (especially in blaze) because now there are two separate channels to pass the configuration into the compiler binary
6. just because we need to produce multiple things doesn't mean we need to run `ngc` multiple times. I'm sure that whatever structure we need can be encoded in tsconfig.json

Also the basePath should not be an argument to the program, it must be the directory containing the tsconfig.json file.
@mhevery can we please revisit?

**Pull request merged**